tests/imap/test_folder_sync.py

```diff
@@ -428,3 +428,67 @@ def test_imap_message_deduplication(
         .count()
         == 1
     )
+
+
+def test_gmail_handle_uidinvalid(
+    db, default_account, all_mail_folder, mock_imapclient, label
+) -> None:
+    """
+    Tests when a GMail account's UID validity changes. In this case we attempt
+    to resync existing messages by their google message ID. If this fails, we
+    delete the message and trust that we'll sync it back in if it still exists.
+
+    When deleting these messages, during the `yield_chunk`, it was possible for
+    a `sqlalchemy.exc.ProgrammingError: (MySQLdb.ProgrammingError)
+    (2014, "Commands out of sync; you can't run this command now")` to be raised.
+
+    This error is raised when an association attempts a cascade delete from
+    within the ORM. Adding label associations triggers the error. To resolve
+    this, we can use `passive_deletes=True` in the `relationship` call.
+    """
+    uid_dict = uids.example()
+    mock_imapclient.add_folder_data(all_mail_folder.name, uid_dict)
+    mock_imapclient.list_folders = lambda: [
+        ((b"\\All", b"\\HasNoChildren"), b"/", "[Gmail]/All Mail")
+    ]
+    mock_imapclient.idle = lambda: None
+    mock_imapclient.idle_check = raise_imap_error
+
+    folder_sync_engine = GmailFolderSyncEngine(
+        default_account.id,
+        default_account.namespace.id,
+        all_mail_folder.name,
+        default_account.email_address,
+        "gmail",
+        BoundedSemaphore(1),
+    )
+    folder_sync_engine.initial_sync()
+
+    # We need to add labels to our imap uids so that the cascade delete on
+    # LabelItems is required. This would trigger a deletion during the yield_chunk
+    # query if any ORM level cascades are triggered.
+    imap_uids = (
```

Comment:
Nit: I think we should add a sanity check assertion that checks the number of `ImapUid` currently stored in the database is what we expect it to be. Currently it looks to me like we trust `initial_sync` to add stuff to the database, but I'd rather tests weren't so trusting.